Added tutorial on HW-accelerated video encoding. (#169)
ignitionrobotics/ign-common: tutorials/hw-encoding.md

> \page hw-encoding HW-accelerated Video Encoding

Reviewer: ```suggestion
\page hw-encoding Hardware accelerated Video Encoding
```

When recording video using the `ignition::common::VideoEncoder` class, you can

> opt-in to use HW acceleration for the encoding process. By default, only software

Reviewer: ```suggestion
opt-in to use hardware (HW) acceleration for the encoding process. By default, only software
```

encoders are used. This tutorial will show how to configure the encoder for HW
acceleration and will present ready-made commandlines for some typical use-cases.

You can either use the `VideoEncoder` class directly, or you can "meet it" in the
video recorder plugin in Ignition Gazebo. In both cases, HW-accelerated encoding
can be set up.

HW acceleration should provide you with higher encoding performance, potentially
leaving more CPU power to the rest of your program/simulation, while taking a bit of
GPU memory (video encoding uses different chips than 3D graphics or CUDA computations,

> so performance-wise, the rest of the GPU should be unaffected).

Author: I added the links to the table of supported encoders. Do you think that's enough?

> so performance-wise, the rest of the GPU should be unaffected).

Reviewer: Yes, I don't think it's something that we need to maintain, as the wikipedia article should stay current and is more informational than anything.

> # Configuring HW acceleration

Reviewer: ```suggestion
# Configuring hardware acceleration
```

Adding support for more is possible and should not be a big problem, but needs
someone with the right hardware to verify the implementation.

The acceleration libraries may also need some drivers for them to work:

- NVEnc: `libnvidia-encode-*` or just CUDA runtime
- VA-API: `libva-glx2`, `libva-drm2`, maybe also `libglvnd0`
- QSV: Unclear, usually works out of the box

## 2. Device names

If your computer has more GPUs, it is important to specify which one to use.
Here are some basic naming rules:

- NVEnc
  - Linux: Devices `/dev/nvidia0`, `/dev/nvidia1` etc. Beware that the numbers 
    in names of these devices can differ from the numbers CUDA assigns the GPUs
    (CUDA orders the GPUs according to their compute capability, whereas numbering
    of these devices is probably by the order on PCI bus).
  - Windows: Devices are called the same as on Linux, even though such files do
    not exist in the system.
- QSV
  - Windows: The devices are called just `0`, `1` and so on.
  - Linux: When using QSV on Linux, you can use VA-API device names.
- VA-API
  - Linux: 
    - DRM: The devices are called `/dev/dri/renderD128`, `/dev/dri/renderD129`

>       etc. DRM stands for Direct Rendering Manager, not Digital Rights Management :)

Reviewer: ```suggestion
      etc. DRM stands for Direct Rendering Manager, not Digital Rights Management
```

      To use these devices, make sure your user has write permissions to the 
      device files. Ubuntu usually doesn't give write access to everybody, but just
      to members of the `video` group.
    - GLX: You can also pass an X-server string like `:0`, which means the encoder
      should use the GPU on which this X server is running (it needs to support
      the GLX extension). On headless machines, you should use DRM.
      
## 3. Using HW surface

The last thing you need to decide is whether the selected encoder should use
a HW surface (pixel buffer) or the default CPU-located one. With most encoders,
this is just a performance issue and they will work both with CPU and GPU surfaces.

It is best if you perform experiments with HW surface used and not used and compare

> the performance. Select the one that suits you better.

Reviewer: ```suggestion
the performance. Select the one that suits your use case better.
```

[GUI] [Msg] Encoder h264_vaapi is not allowed, skipping.
[GUI] [Msg] Found known HW encoder: nvenc
[GUI] [Err] [Util.cc:58] ffmpeg [AVHWDeviceContext] Could not initialize the CUDA driver API

[GUI] [Err] [HWEncoder.cc:159] Could not initialize device /dev/nvidia3 for NVENC, skipping.
[GUI] [Msg] Found known HW encoder: nvenc_h264
[GUI] [Err] [Util.cc:58] ffmpeg [AVHWDeviceContext] Could not initialize the CUDA driver API

[GUI] [Err] [HWEncoder.cc:159] Could not initialize device /dev/nvidia3 for NVENC, skipping.
[GUI] [Wrn] [HWEncoder.cc:384] No hardware-accelerated encoder found, falling back to software encoders
[GUI] [Msg] Compatible SW encoder: libx264
[GUI] [Msg] Using encoder libx264
```

# Examples

Here are a few ready-made examples which might or might not work for you right-away.
Just give them a try and dig deeper in the configuration if something is wrong.

## Linux/Win + Intel GPU

    IGN_VIDEO_ALLOWED_ENCODERS=QSV

## Linux/Win + NVidia GPU

    IGN_VIDEO_ALLOWED_ENCODERS=NVENC

## Linux + Intel/NVidia GPU

    IGN_VIDEO_ALLOWED_ENCODERS=VAAPI

## Linux NVidia Multi-GPU machine

    IGN_VIDEO_ALLOWED_ENCODERS=NVENC IGN_VIDEO_ENCODER_DEVICE=/dev/nvidia2

> # Pitfalls

Reviewer: ```suggestion
# Caveats
```

## NVEnc per-machine limit

If you have a multi-GPU station with desktop-class (not server-class) GPUs, you will
run into an artificial limitation from NVidia. You can only run 3 concurrent
encoding sessions on one computer, no matter how many GPUs you have. The computer
can even have 8 GPUs, but you will only be able to encode 3 videos at a time. The
exact maxima of encoding sessions are described in
[this NVidia support page](https://developer.nvidia.com/video-encode-and-decode-gpu-support-matrix-new).

To make things worse, there is no API that would tell you the number of currently
running NVEnc sessions. The only way to find out you're launching the fourth is
trying to start encoding and getting a memory allocation error. This library
catches this error and writes a lengthy description of what might have just happened
(either really low memory or this artificial limit). Unfortunately, when you start
the doomed fourth encoding session, all the three "legal" sessions will crash.
This might be really troublesome on e.g. multi-user systems when you don't even know
which jobs of the other users are using NVEnc.

> There is a workaround removing this artificial limit - patching the binary blob
> drivers using https://github.com/keylase/nvidia-patch . It is up to you if you
> can and want to do that. And there is no guarantee it will work forever.

Reviewer: ```suggestion
There is a workaround removing this artificial limit - patching the binary blob
drivers using https://github.com/keylase/nvidia-patch . This is an unofficial 
patch that is not supported by NVidia or the Ignition developers. It is up to 
you if you can and want to do that. And there is no guarantee it will work forever.
```
